### ISSUE TITLE
missing numpy functions

### ISSUE THREAD

**Issue by haoyuan80s** [emacs-lsp/lsp-python-ms#101]
Hi,

Seems that it misses many numpy functions.
![image](https://user-images.githubusercontent.com/9064146/83346497-c24e5700-a2d1-11ea-8804-62c6ce92f8a7.png)
Any way to fix it?

**Comment by seagle0128** [comment]
It works for me. No details, and not able to troubleshoot.

![image](https://user-images.githubusercontent.com/140797/83349190-7cdb6980-a365-11ea-91d1-3b6e48d90265.png)

**Comment by seagle0128** [comment]
I guess it's related to `company` and `company-lsp`. Recommend to use `lsp-prefer-capf`.

**Comment by haoyuan80s** [comment]
Can you confirm that you have `np.random.rand` function in the company candidates?

Here the the results using `pyls`.
![image](https://user-images.githubusercontent.com/9064146/83356801-4b3eb000-a31d-11ea-9b52-532d6d371ef7.png)

**Comment by seagle0128** [comment]
Can you try with VSCode to check if it's same result?
I found this https://github.com/microsoft/python-language-server/issues/1847.

**Comment by haoyuan80s** [comment]
I also noticed their discussion.

But my VSCode can show `np.random.rand`. (Not show which language server are they using thought.)

<img width="372" alt="image" src="https://user-images.githubusercontent.com/9064146/83435860-c5d8ff80-a3f1-11ea-9e8d-d9a901dbfffe.png">

And seems that VSCode can show the results much faster.

**Comment by seagle0128** [comment]
I believe you are using different version of ms pyls in vscode.

**Comment by seagle0128** [comment]
Someone mentioned "However this doesn't happen when I use virtualenv environment".
See https://github.com/microsoft/python-language-server/issues/2043

**Comment by haoyuan80s** [comment]
Yes, seems that the python virtual environment is problem here.  I have the issue in a virtual environment built by command `python -m venv evn_name`. I just tested it in a python without any virtual environment, there is no such issue.

<img width="372" alt="image" src="https://user-images.githubusercontent.com/9064146/83605305-86fb9480-a52c-11ea-8e28-51cd5ab72c84.png">

**Comment by seagle0128** [comment]
So it's an issue of ms pyls, not lsp-python-ms, right?

**Comment by haoyuan80s** [comment]
Yes. It should some bug of ms pyls. 

The problem is not from virtual environment. It should be package related.  If I just install `numpy` in a virtual env, it can complete successfully. But after intall `torch`, then some `numpy` functions are missing.

**Comment by seagle0128** [comment]
Please report to ms pyls, then. I am closing it.

**Comment by anupam312nwd** [comment]
> Yes. It should some bug of ms pyls.
> 
> The problem is not from virtual environment. It should be package related. If I just install `numpy` in a virtual env, it can complete successfully. But after intall `torch`, then some `numpy` functions are missing.

@haoyuan80s 
Hi, is auto-completion for numpy working for you now? For me, auto-completion is not working properly for "re" python library  and "torch" library (along with "numpy") too? Auto-completion for "numpy" and "re" are working in company-mode and in VSCode fine.
I don't know enough about these to open a new issue. I couldn't find any similar post in "mspyls" git repository.

**Comment by seagle0128** [comment]
@anupam312nwd I recommend to try [lsp-pyright](https://github.com/emacs-lsp/lsp-pyright/).

**Comment by anupam312nwd** [comment]
> @anupam312nwd I recommend to try [lsp-pyright](https://github.com/emacs-lsp/lsp-pyright/).

Thanks. I tried pyls (I think https://github.com/palantir/python-language-server) and it's working better than mspyls. I will later try lsp-pyright too.

**Comment by haoyuan80s** [comment]
@anupam312nwd I am using pyls + lsp-pyright(very useful). It works well for me.

**Comment by haoyuan80s** [comment]
@seagle0128 Lsp-pyright also misses some numpy/pytorch functions.
![image](https://user-images.githubusercontent.com/9064146/93691098-7c421c80-fa95-11ea-8c72-6518a9be84a4.png)

Any hint how to fix it?

**Comment by anupam312nwd** [comment]
> @seagle0128 Lsp-pyright also misses some numpy/pytorch functions.
> ![image](https://user-images.githubusercontent.com/9064146/93691098-7c421c80-fa95-11ea-8c72-6518a9be84a4.png)
> 
> Any hint how to fix it?

yes. I got the same thing with lsp-pyright. In doom-emacs (I am using doom), setting lsp using `init` file gives 3 mutually exclusive option: mspyls (default), pyls and py-right. I used `(package! lsp-python-ms :disable t)` to disable mspyls. `pyls` seems to me best choice so far. I found this comment https://github.com/hlissner/doom-emacs/issues/2600#issuecomment-590938993 useful.
